File: hier/project-euler/euler-003-hackerrank/euler003.java
import java.util.Scanner;
import java.util.stream.IntStream;
import java.util.stream.LongStream;

final class Arith {
    static int sqrt32(final int n)
    {
        return (int)Math.round(Math.floor(Math.sqrt(n)));
    }
    
    static long sqrt64(final long n)
    {
        return Math.round(Math.floor(Math.sqrt(n)));
    }
    
    static int sqrt64to32(final long n)
    {
        return (int)sqrt64(n);
    }
    
    private static IntStream rangeClosed(
            final int startInclusive, final int endInclusive, final int step)
    {
        return IntStream.rangeClosed(0, (endInclusive - startInclusive) / step)
                        .map(k -> startInclusive + k * step);
    }
    
    static LongStream sieve(final int n) // streams primes up to n
    {
        final boolean[] a = new boolean[n + 1];
        
        final int imax = sqrt32(n);
        for (int i = 3; i <= imax; i += 2) {
            if (!a[i]) {
                final int jstep = i * 2;
                for (int j = i * i; j <= n; j += jstep) a[j] = true;
            }
        }
        
        return LongStream.concat(LongStream.of(2L),
                rangeClosed(3, n, 2).filter(k -> !a[k]).asLongStream());
    }
    
    // Computes the largest prime factor of n > 1, provided primes contains at
    // least all prime numbers <= sqrt(n) and no non-prime numbers.
    static long maxPrimeDivisor(final long[] primes, long n)
    {
        while ((n & 1L) == 0L) n >>>= 1L; // optimization, removes factors of 2
        if (n == 1L) return 2L;
        
        final long sqrt = sqrt64(n);
        
        for (int i = 1; i != primes.length && primes[i] <= sqrt; ++i) {
            if (n % primes[i] != 0L) continue;
            
            do n /= primes[i];
            while (n % primes[i] == 0L);
            
            if (n == 1L) return primes[i];
        }
        
        return n; // what remains must be a single prime factor > sqrt(orig. n)
    }
    
    private Arith() { } // the Arith class should not be instantiated
}

final class Solution {
    static long[] readValues()
    {
        try (final Scanner sc = new Scanner(System.in)) {
            return IntStream.range(0, sc.nextInt())
                            .mapToLong(i -> sc.nextLong())
                            .toArray();
        }
    }
    
    public static void main(final String[] args)
    {
        final long[] values = readValues();
        final long max = LongStream.of(values).max().getAsLong();
        
        final long[] primes = Arith.sieve(Arith.sqrt64to32(max)).toArray();
        
        for (final long n : values)
            System.out.println(Arith.maxPrimeDivisor(primes, n));
    }
    
    private Solution() { } // the Solution class should not be instantiated
}

File: hier/project-euler/euler-003-hackerrank/euler003.vb
Option Explicit On
Option Strict On

Imports System.Collections.Generic
Imports System.Linq
Imports System.Runtime.CompilerServices

<Assembly: Reflection.AssemblyVersion("1.0.0.0")>

Friend Module Arith
    <Extension()>
    Friend Function Sqrt32(n As Integer) As Integer
        Return CInt(Math.Floor(Math.Sqrt(n)))
    End Function

    <Extension()>
    Friend Function Sqrt64(n As Long) As Long
        Return CLng(Math.Floor(Math.Sqrt(n)))
    End Function

    <Extension()>
    Friend Function Sqrt64To32(n As Long) As Integer
        Return CInt(Math.Floor(Math.Sqrt(n)))
    End Function

    Private Function RawSieve(n As Integer) As Boolean() ' finds primes to n
        Dim a(n) As Boolean ' false means not marked as composite

        For i = 3 To Sqrt32(n) Step 2
            If Not a(i) Then
                For j = i * i To n Step i * 2
                    a(j) = True
                Next
            End If
        Next

        Return a
    End Function

    Private Iterator Function Sieve(n As Integer) As IEnumerable(Of Long)
        Dim a = RawSieve(n)

        Yield 2L
        For k = 3 To n Step 2
            If Not a(k) Then Yield k ' int-to-long conversion happens here
        Next
    End Function

    <Extension()>
    Friend Function Primes(n As Integer) As Long() ' primes to n, as array
        Return Sieve(n).ToArray()
    End Function

    <Extension()>
    Friend Function MaxPrimeDivisorOf(primes As IEnumerable(Of Long),
                                      n As Long) As Long
        While (n And 1L) = 0L ' optimization to remove factors of 2
            n >>= 1
        End While
        If n = 1L Then Return 2L

        Dim sqrt = Sqrt64(n)
        For Each p In primes.Skip(1).TakeWhile(Function(q) q <= sqrt)
            If (n Mod p = 0L) Then
                Do
                    n \= p
                Loop While n Mod p = 0L

                If n = 1L Then Return p
            End If
        Next

        Return n ' what remains must be a single prime factor > sqrt(orig. n)
    End Function
End Module

Friend Module Solution
    Private Function ReadValues() As List(Of Long)
        Dim t = Integer.Parse(Console.ReadLine)
        ReadValues = New List(Of Long)(t)

        For i = 1 To t
            ReadValues.Add(Long.Parse(Console.ReadLine))
        Next
    End Function

    Friend Sub Main()
        Dim values = ReadValues()
        Dim primes = values.Max.Sqrt64To32.Primes

        values.ForEach(Sub(n) Console.WriteLine(primes.MaxPrimeDivisorOf(n)))
    End Sub
End Module
